Feat: Add http/http keyword timeout option

zenyr opened the pull request:
Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Description
- Adds `timeoutMs` to monitor and use it
  - I added 'ms' to make sure the value is NOT a *second*, unlike intervals
- Adds `timeout_ms` to sql patch list
- Fixes #2140, Fixes #2077, Fixes #877

## Type of change

- User interface (UI)
- New feature (non-breaking change which adds functionality)
- Translation update (minor and 🇰🇷  only, running script added a bunch of missing i18n data 😏  )
  -   edit: reverted non-english & non-Korean files

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings

## Screenshots (if any)

<img width="1032" alt="image" src="https://user-images.githubusercontent.com/4316558/193277303-4e5caefb-e9f9-4bc8-ac19-add8c7eebe73.png">

Computroniks reviewed (changes requested):
Please don't add the English translations to non English files as it will fall back to the English translation if no other translation is found and will just cause merge conflicts later. I assume you ran `npm run update-language-files`.

@louislam Should we perhaps change this script so it doesn't update all the files and instead only changes a specific one? I don't really see any situation where it it useful and it only seems to create work for contributers when they run it and commit the results, only to be told to revert it.

zenyr commented:
I see, let me update the PR to comply the best practice!

zenyr commented on src/languages/ko-KR.js:
```diff
@@ -520,12 +520,62 @@ export default {
+    RadiusSecret: "Radius 비밀 정보",
+    RadiusSecretDescription: "클라이언트와 서버간 공유 비밀 키",
+    RadiusCalledStationId: "호출된 스테이션 아이디",
+    RadiusCalledStationIdDescription: "호출된 기기 식별자",
+    RadiusCallingStationId: "호출한 스테이션 아이디",
+    RadiusCallingStationIdDescription: "호출하는 기기 식별자",
+    "Connection String": "연결 문자열",
+    Query: "쿼리",
+    settingsCertificateExpiry: "TLS 인증서 만료",
+    certificationExpiryDescription: "TLS 인증서가 다음 안에 만료되는 경우 HTTPS 모니터가 알림을 보냅니다:",
+    "Setup Docker Host": "Docker 호스트 설정",
+    "Connection Type": "연결 형태",
+    "Docker Daemon": "Docker 데몬",
+    deleteDockerHostMsg: "모든 모니터에서 이 Docker 호스트를 삭제하시겠습니까?",
+    socket: "소켓",
+    tcp: "TCP / HTTP",
+    "Docker Container": "Docker 컨테이너",
+    "Container Name / ID": "컨테이너명 / ID",
+    "Docker Host": "Docker 호스트",
+    "Docker Hosts": "Docker 호스트들",
+    "ntfy Topic": "ntfy 토픽",
+    Domain: "도메인",
+    Workstation: "워크스테이션",
+    disableCloudflaredNoAuthMsg: "현재 비인증 모드이므로, 비밀번호가 필요하지 않습니다.",
+    wayToGetLineNotifyToken: "{0}에서 액세스 토큰을 얻을 수 있습니다.",
+    Examples: "예제",
+    "Home Assistant URL": "홈 어시스턴트 URL",
+    "Long-Lived Access Token": "장기 액세스 토큰",
+    "Notification Service": "알림 서비스",
+    "default: notify all devices": "기본: 모든 기기에 알림",
+    "Automations can optionally be triggered in Home Assistant:": "자동화를 홈 어시스턴트에서 선택적으로 동작시킬 수 있습니다.:",
+    "Trigger type:": "트리거 유형:",
+    "Event type:": "이벤트 유형:",
+    "Event data:": "이벤트 데이터:",
+    "Then choose an action, for example switch the scene to where an RGB light is red.": "이후 액션을 선택하세요, 예를 들어 RGB 불빛이 적색인 Scene 등으로 교체합니다.",
+    "Frontend Version": "프론트엔드 버전",
+    "Frontend Version do not match backend version!": "프론트엔드 버전이 백엔드 버전과 맞지 않습니다!",
+    "Base URL": "기본 URL",
+    goAlert: "GoAlert",
+    backupRecommend: "볼륨 또는 데이터 폴더 (./data/) 를 대신 직접 백업해주세요.",
```
Hello team, I've reverted every un-touched languages change. Please take a look again and see if there's something I could improve. In the meantime I may take a look into the script itself too but I'll refrain myself from mixing 2 features into 1 PR 😄

zenyr commented:
I forgot to commit the updated en.js, just pushed one! 😼

Computroniks commented on src/pages/EditMonitor.vue:
```diff
@@ -606,6 +611,10 @@ export default {
                 this.monitor.retryInterval = value;
             }
+            // keep timeoutMs below interval, minding the unit conversion
+            if (!this.monitor.timeoutMs || this.monitor.timeoutMs >= value * 1000) {
```
It might be a good idea to have some sort of limit to make sure that the previous request has definitely finished before the next one is sent

zenyr replied:
If you prefer `less or equal than 80% of the interval seconds` I can change it at once ;) 
Please let me know. (cc. @louislam asking for owner's opinion)

zenyr commented:
> > Please don't add the English translations to non English files as it will fall back to the English translation if no other translation is found and will just cause merge conflicts later. I assume you ran `npm run update-language-files`.
> > @louislam Should we perhaps change this script so it doesn't update all the files and instead only changes a specific one? I don't really see any situation where it it useful and it only seems to create work for contributers when they run it and commit the results, only to be told to revert it.
> 
> Oh yes, the command is bad.

Yeah, so I reverted all the artifacts that it created in this PR 😄 Since my team needs this feature and fixes a lot of pending issues is there something I could do to proceed with this PR? I'm sorry if I'm too attached to this feature but due to the fact our VPS network being flakey (like, losing a packet or few randomly) I really would like to use this feature in the stable version..! 😅 

Thank you and ❤️ your work, 
UK made me brave enough to hop into Vue train even though I've been a long time ⚛️ enthusiast.

Eliepse commented:
Would love to see this feature merged. Is there anything that prevent it? I'm willing to help if I can :)

Eliepse commented:
Just tested this PR, works fine for me 👍 

![image](https://user-images.githubusercontent.com/3075627/204082951-fff52a61-fa98-456f-96ba-c91d7dbe5c7b.png)

usufu commented:
@zenyr any docker image for this feature? or how to update docker image with this feature? I need this feature. thx.
why the pull request still not merged? @Computroniks @louislam

i-am-dev commented:
Hi.

Any update on this PR and possibly when it will be merged? 

Thanks.

i-am-dev commented:
> Hi.
> 
> Any update on this PR and possibly when it will be merged?
> 
> Thanks.

Nevermind. I see now that it is part of the 1.20.0 milestone.

Any due date for the release yet?

Thanks.

louislam commented:
Going to work on it. I merged conflict and added `v-if` to the field, so it shows only if http(s) related monitors.

But there are some questions/issues:
- Timeout value must be within the check interval
- As other time related is using second as the unit, I think it should use second instead of millisecond
- I feel `Heartbeat Timeout` is a bit similar to `Heartbeat Interval`, which may be a bit confusing. How about rename it to `Request Timeout`? (If any native English native speakers here, feel free to comment)

lightingman117 commented:
USA - Request Timeout is better.

At first, I had no idea what this feature was doing since there is already a timeout notification.

I see that it now allows editing that default value (48000ms) on a per-monitor basis.

chakflying commented:
Maybe it would be better to put this in advanced since it might be confusing? We can also have a short description below for more detailed explanation.

usufu commented:
can anyone help merge the code? there are some conflicts must be resolved.

madan-wego commented:
can anyone merge this code ? needed this changes

jeloneal commented:
I also want to upvote this one, will be very handy once available. Thanks!

usufu commented:
@louislam pls take time to merge this feature, thx.

fredericpellin commented:
it could be great if it can be merge for the next release ;) @louislam

usufu commented:
@louislam

arthur-lbchn commented:
do you have any ETA to have this merged? thanks!

slimshizn commented:
Yes please merge. It has been quite a while.

zenyr commented:
Thank you for your assistance in resolving the conflicts and all the interests. I have matched them once again, and I would appreciate it if you could review my work when you get the chance! ❤️

kdckrs commented:
Awesome! Will be happy when this feature is released so I can finally shut down my other monitoring system.

louislam commented:
Hello @zenyr, my last comment is here.

> Going to work on it. I merged conflict and added `v-if` to the field, so it shows only if http(s) related monitors.
> 
> But there are some questions/issues:
> 
>     * Timeout value must be within the check interval
> 
>     * As other time related fields are using second as the unit, I think it should use second instead of millisecond
> 
>     * I feel `Heartbeat Timeout` is a bit similar to `Heartbeat Interval`, which may be a bit confusing. How about rename it to `Request Timeout`? (If any native English native speakers here, feel free to comment)

Also I think @chakflying's concern is reasonable.

> Maybe it would be better to put this in advanced since it might be confusing? We can also have a short description below for more detailed explanation.

zenyr commented:
> Hello @zenyr, my last comment is here. [#2142 (comment)](https://github.com/louislam/uptime-kuma/pull/2142#issuecomment-1408729916)
> 
> > Going to work on it. I merged conflict and added `v-if` to the field, so it shows only if http(s) related monitors.
> > But there are some questions/issues:
> > ```
> > * Timeout value must be within the check interval
> > 
> > * As other time related fields are using second as the unit, I think it should use second instead of millisecond
> > 
> > * I feel `Heartbeat Timeout` is a bit similar to `Heartbeat Interval`, which may be a bit confusing. How about rename it to `Request Timeout`? (If any native English native speakers here, feel free to comment)
> > ```
> 
> Also I think @chakflying's concern is reasonable. [#2142 (comment)](https://github.com/louislam/uptime-kuma/pull/2142#issuecomment-1414228676)
> 
> > Maybe it would be better to put this in advanced since it might be confusing? We can also have a short description below for more detailed explanation.
> 
> And @lightingman117 suggested a better name for it:
> 
> [#2142 (comment)](https://github.com/louislam/uptime-kuma/pull/2142#issuecomment-1413942615)

I understand and agree with your points. While setting a timeout longer than the interval may seem.. shooting their own foot 😅 , finding the right balance is indeed crucial.

Regarding the timeout resolution, it's common to work with whole numbers in seconds for consistency. However, if there is a requirement for sub-second resolution, it may be necessary to use floating-point values to accommodate those needs. The choice between using single or float would depend on your choice!

borgogelli commented:
> already a timeout notification

@lightingman117 you have written: "there is already a timeout notification"
But where is it ?

louislam commented:
The edit page is no longer working, here is the browser console log:

![image](https://github.com/louislam/uptime-kuma/assets/1336778/d475b98b-8246-4362-9462-c94a06e6dc31)

zenyr commented on src/pages/EditMonitor.vue:
```diff
@@ -1201,6 +1219,14 @@ message HealthCheckResponse {
+            // limit to 80% of interval
+            const maxTimeout = this.monitor.interval * 0.8;
+            // 0 will be treated as 80% of interval
+            this.monitor.timeout = Math.Max(0, Math.min(timeout, maxTimeout));
```
This is a very amateur like mistake from my side. My bad!

zenyr commented:
Just resolved "Math.Max" typo for the PR. Sorry for the hassle! I must've been too excited when I added the line!

louislam commented:
Test it again, unfortunately, newly created monitor doesn't seem to be working. 

Weird default value:
![image](https://github.com/louislam/uptime-kuma/assets/1336778/fd40c2f2-b7b0-4e88-9b33-119bcd83d9b8)

I manually inputted 48 seconds, but the monitor timeout in 48ms. Forgot to x1000?
![image](https://github.com/louislam/uptime-kuma/assets/1336778/2709a8e9-6e55-4aa9-8880-879f6ab0d442)

Old monitor, the original interval is 60s and the request time is 48s, then I change the interval to 30s.

The request time became this:
![image](https://github.com/louislam/uptime-kuma/assets/1336778/3bb8f49b-7067-4f3e-8a4f-209436b6ee6f)

zenyr commented:
Sure! I will ping you again once I finish testing the code first, I had issues setting up dev env on a new device(for personal reasons) but definitely will do!

zenyr commented:
Fixed few things
1. Co-located timeout field with intervals, making their relation more clear
2. Used different formula to clamp to 80%, resulting better fraction numbers for edit/add view. ( `t*0.8` => `~~(t * 8)/10`)
3. Displays 80% inside the UI that directly depicts `0 == default value(80%)` (since timeout of 0 literally makes 0 sense)
  - This is not visible in the GIF below but in the source code, `monitor.timeout || clampTimeout( interval )` is used.
![timeout-test](https://github.com/louislam/uptime-kuma/assets/4316558/29ea6a72-4e81-4a45-bbb5-64a1109aec2d)
4. Added db migration code to the last place

CommanderStorm commented on src/pages/EditMonitor.vue:
```diff
@@ -1120,6 +1134,7 @@ message HealthCheckResponse {
                     url: "https://",
                     method: "GET",
                     interval: 60,
+                    timeout: this.interval * 0.8, // previous default value
```
Previous compared to what?
I think this comment will not age well ^^

```suggestion
                    timeout: this.interval * 0.8,
```
